src/lib/openjp2/segmented_stream.c

```diff
@@ -0,0 +1,335 @@
+/*
+ * The copyright in this software is being made available under the 2-clauses 
+ * BSD License, included below. This software may be subject to other third 
+ * party and contributor rights, including patent rights, and no such rights
+ * are granted under this license.
+ *
+ * Copyright (c) 2015, Aaron Boxer
+ * All rights reserved.
+ *
+ * Redistribution and use in source and binary forms, with or without
+ * modification, are permitted provided that the following conditions
+ * are met:
+ * 1. Redistributions of source code must retain the above copyright
+ *    notice, this list of conditions and the following disclaimer.
+ * 2. Redistributions in binary form must reproduce the above copyright
+ *    notice, this list of conditions and the following disclaimer in the
+ *    documentation and/or other materials provided with the distribution.
+ *
+ * THIS SOFTWARE IS PROVIDED BY THE COPYRIGHT HOLDERS AND CONTRIBUTORS `AS IS'
+ * AND ANY EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT LIMITED TO, THE
+ * IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS FOR A PARTICULAR PURPOSE
+ * ARE DISCLAIMED.  IN NO EVENT SHALL THE COPYRIGHT OWNER OR CONTRIBUTORS BE
+ * LIABLE FOR ANY DIRECT, INDIRECT, INCIDENTAL, SPECIAL, EXEMPLARY, OR
+ * CONSEQUENTIAL DAMAGES (INCLUDING, BUT NOT LIMITED TO, PROCUREMENT OF
+ * SUBSTITUTE GOODS OR SERVICES; LOSS OF USE, DATA, OR PROFITS; OR BUSINESS
+ * INTERRUPTION) HOWEVER CAUSED AND ON ANY THEORY OF LIABILITY, WHETHER IN
+ * CONTRACT, STRICT LIABILITY, OR TORT (INCLUDING NEGLIGENCE OR OTHERWISE)
+ * ARISING IN ANY WAY OUT OF THE USE OF THIS SOFTWARE, EVEN IF ADVISED OF THE
+ * POSSIBILITY OF SUCH DAMAGE.
+ */
+
+#include "opj_includes.h"
+
+/*--------------------------------------------------------------------------*/
+
+/*  Segmented Buffer Stream */
+
+static void opj_seg_buf_increment(opj_seg_buf_t * seg_buf)
+{
+	opj_buffer_t* cur_seg = NULL;
+	if (seg_buf == NULL ||	seg_buf->cur_seg_id == seg_buf->segments.size-1) {
+		return;
+	}
+
+	cur_seg = (opj_buffer_t*)opj_vec_get(&seg_buf->segments, seg_buf->cur_seg_id);
+	if (cur_seg->offset == cur_seg->len  && 
+					seg_buf->cur_seg_id < seg_buf->segments.size -1)	{
+		seg_buf->cur_seg_id++;
+	}
+}
+
+static OPJ_SIZE_T opj_seg_buf_read(opj_seg_buf_t * seg_buf,
+									void * p_buffer,
+									OPJ_SIZE_T p_nb_bytes)
+{
+	OPJ_SIZE_T bytes_in_current_segment;
+	OPJ_SIZE_T bytes_to_read;
+	OPJ_SIZE_T total_bytes_read;
+	OPJ_SIZE_T bytes_left_to_read;
+	OPJ_SIZE_T bytes_remaining_in_file;
+
+	if (p_buffer == NULL || p_nb_bytes == 0 || seg_buf == NULL)
+		return 0;
+
+	/*don't try to read more bytes than are available */
+	bytes_remaining_in_file = seg_buf->data_len - opj_seg_buf_get_global_offset(seg_buf);
+	if (p_nb_bytes > bytes_remaining_in_file) {
+		printf("Warning: attempt to read past end of segmented buffer\n");
```

Comment:
There should be no printf in the library code. Please use the existing message mechanism if necessary.